design-decisions.md

```diff
@@ -0,0 +1,28 @@
+# Design Decisions
+Assumptions, thoughts and decisions for current and future consideration.
+
+##  Technologies and frameworks
+- Typescript
+- TSLint
+- DynamoDb - The problem prefers it and it has good options for local development in a docker image or serverless plugin
+- Documentation - Rely on github for now but it would be nice to have polished api documentation
+- Serverless provisioning (serverless, AWS Sam, Terraform...) - I'd like to try out AWS Sam but I'm familiar with serverless so I'll choose it for this project
+- Local dev environment - I would normally begin with a fully functioning local dev environment (serverless-offline, serverless-dynamodb-local) - This is going to take a little more time but I still think it's essential to start any real project this way.  I'll invest the extra time on this.
+- serverless-dynamodb-local - We have other options with dynamodb-local through [docker](https://hub.docker.com/r/amazon/dynamodb-local) which I've used recently.  But I really wanted to evaluate this library. Unfortunately it requires Java.  For me that's not a huge issue but something that could become an issue for larger teams.
+
+- Validation - Try to handle with Json Shema and API Gateway.  I was able to get validation working with API Gateway and a schema defined in serverless BUT there is an open issue in serverless-offline which means it will not work locally.  Thus I wasn't able to add tests for it.  See below for a more optimal long-term approach.  
+- Testing (mocha, tsnode, jest) - I chose jest with ts-jest.  It worked well enough.  I ran into an issue where my unit tests really weren't going to offer much value but integration tests would. Integration tests working with all the local tooling to replicate the aws serverless environment with a single command would certainly be possible and optimal but was taking too much time.  For now I will run two commands to run tests :(
+
+## Future Considerations
+- Logging - I'm using pino which I've used before.  It's fast and defaults to json which works well with cloudwatch and other logging aggregators.  We can enhance logging to record information about each request.  We could then combine logging with cloudwatch to cheaply get some traction on a few items below (monitoring and analytics).
+- Monitoring - We would want to introduce monitoring and alarms for latency, call volume, errors etc
+- Security - I have used AWS WAF to protect against DDoS, general bots, throttling, etc in the past.
+- Analytics - This may or may not be important.  Something to consider while evaluating the needs of the client
```

Comment:
> When it has been necessary, can you give us an example of one or two tools you have used for analytics on domain operations within the service?

@afaq-sf Sure, at Onu I used Keen.io to record usage stats.  I added a layer into my rest framework's request/response chain.  It would record certain events.  For example, we had a client who wanted to know which sales persons were logging in a which subset of API calls they were making and how often.  So we could record this to Keen and then we would query the data back for reporting.  Similarly we would record total data transferred to/from our media service by upstream services and their users.  This allowed us to tier pricing if needed for users who had heavy media usage.  

In general (before Keen) I also used a cloudwatch reporter for our java rest framework (dropwizard).  This allowed us to easily record request/response metrics to cloudwatch.  We used it mostly for monitoring but it could easily be used for some short term analytics.